byron/ledger/executable-spec/src/Ledger/UTxO/Generators.hs

```diff
@@ -0,0 +1,273 @@
+    . bitraverse runTreeT runTreeT
+
+interleaveInputOutput
+  :: Monad m
+  => (input -> Integer)
+  -> (output -> Integer)
+  -> ((Integer -> Integer) -> output -> output)
+  -> (NodeT m [input], NodeT m [output])
+  -> NodeT m ([input], [output])
+interleaveInputOutput inValue outValue modifyOutValue (as, bs) =
+  case nodeValue as of
+    [] -> NodeT ([], []) []
+    _  -> NodeT
+      (nodeValue as, nodeValue bs)
+      (  shrinkLeftPreserving inValue outValue modifyOutValue (as, bs)
+      ++ shrinkRight inValue outValue modifyOutValue (as, bs)
+      )
+
+shrinkRight
+  :: Monad m
+  => (input -> Integer)
+  -> (output -> Integer)
+  -> ((Integer -> Integer) -> output -> output)
+  -> (NodeT m [input], NodeT m [output])
+  -> [TreeT m ([input], [output])]
+shrinkRight inValue outValue modifyOutValue (xs, ys1) = do
+  ys2 <- nodeChildren ys1
+  pure . TreeT $ do
+    ys3 <- runTreeT ys2
+    pure $ interleaveInputOutput inValue outValue modifyOutValue (xs, ys3)
+
+shrinkLeftPreserving
+  :: Monad m
+  => (input -> Integer)
+  -> (output -> Integer)
+  -> ((Integer -> Integer) -> output -> output)
+  -> (NodeT m [input], NodeT m [output])
+  -> [TreeT m ([input], [output])]
+shrinkLeftPreserving inValue outValue modifyOutValue (xs1, ys1) = do
+  xs2 <- nodeChildren xs1
+  pure . TreeT $ do
+    xs3 <- runTreeT xs2
+    let
+      lost = sum (inValue <$> nodeValue xs1) - sum (inValue <$> nodeValue xs3)
+      ys2  = subFromList lost outValue modifyOutValue <$> ys1
+    pure $ interleaveInputOutput inValue outValue modifyOutValue (xs3, ys2)
+ where
+  subFromList
+    :: Integer
+    -> (a -> Integer)
+    -> ((Integer -> Integer) -> a -> a)
+    -> [a]
+    -> [a]
+  subFromList n getVal modifyVal = go n
+   where
+    go 0  x        = x
+    go _  []       = []
+    go n' (x : xs) = if getVal x > n'
+      then modifyVal (subtract n) x : xs
+      else go (n' - getVal x) xs
+
+
+-- | A property to check that `genSplitValue` does indeed preserve the input
+propGenSplitValue :: Property
+propGenSplitValue = property $ do
+  n    <- forAll $ Gen.integral (Range.linear 1 10000)
+  ints <- forAll $ genSplitValue n (pure id) id id
+  sum ints === n
+
+-- | Given an input value and functions to generate, view, and update some type
+--   'a' based on that value, split the input into a number of 'a's, preserving
+--   the value throughout the shrink tree
+genSplitValue
+  :: Integer
+  -> Gen (Integer -> a)
+  -> (a -> Integer)
+  -> ((Integer -> Integer) -> a -> a)
+  -> Gen [a]
+genSplitValue n genA getValue modifyValue =
+  mapGenT
+      ( TreeT
+      . (interleaveTreeTPreserving getValue modifyValue . nodeValue =<<)
+      . runTreeT
+      )
+    $ go n []
+ where
+  go 0    acc = pure acc
+  go left acc = do
+    mTree <- toTreeMaybeT (genA <*> Gen.integral (Range.constant 1 left))
+    case runDiscardEffect mTree of
+      Nothing   -> empty
+      Just tree -> go (left - getValue (treeValue tree)) (mTree : acc)
+
+interleaveTreeTPreserving
```

Comment:
It'd be nice to have some comments on what these `*Preserving` functions do. For instance, what do they preserver? The money value? Some invariant?

Reply:
Yeah, good idea